After quality filtering and Illumina sequencing adaptor trimming with FASTX-Toolkit (v0.0.13) , the raw paired-end reads were merged to single-end reads by using FLASh software (v1.2.11, T. Magoc et al., 2011). The correlated 5'-end and 3'-end sequences were extracted by the custom script ('fasta_to_paired.sh') utilizing the SeqKit (v2.4.0) and Cutadapt (v4.1) packages. The inferred full-length reads were generated by Bedtools (v2.31.0, Quinlan et al., 2010) and Samtools (v1.17, Danecek et al., 2021) after mapping to the reference genome (NC000913.3 for Eco, NC008596.1 for Msm, and NC018143.2 for Mtb) via Bowtie2 (v2.5.1, Langmead et al., 2012). The full-length reads with an insert length greater than 10,000 nt were discarded. The mapping results were visualized using the  IGV genome viewer (v2.4.10, Robinson et al., 2011). SEnd-seq coverage was calculated per nucleotide position in the genome, and further analyses were performed using Perl (v5.34.1) or Python (3.11.3) with custom scripts available on Github (https://github.com/LiuLab-codes/Mtb_transcriptome_profiling). Data analysis and visualization scripts utilized Python packages including Matplotlib (v3.7.1), Numpy (v1.24.3), Scipy (v1.10.1), bioinfokit (v0.3), and pyCircos (v0.3.0). Motif analysis was conducted using the MEME suite (v5.5.2, Bailey et al., 2015). The Gene Ontology analysis was performed on DAVID website (v2023q2, Sherman et al., 2022).
